File: Python/counting_sort.py
A = [5, 10, 5, 2, 5, 9, 2, 6, 5, 3, 2]
m = max(A)
count = [0] * (m + 1)

n = len(A)
# print("Len: ", n)
print("Input: ", A)

# print(count)
ops = 0;
for n in A:
    count[n] += 1
    ops += 1

# print(count)

A =[]
for i in range(m + 1):
    A += [i] * count[i]
    ops += 1

print("No of Operations: ", ops)
print("Output: ",A)

File: C++/counting_sort.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int countingSort(vector<int>& arr) {

    int maxVal = *max_element(arr.begin(), arr.end());
    int minVal = *min_element(arr.begin(), arr.end());
    int arr_size = size(arr);
    int range = maxVal - minVal + 1;
    int ops = 0;
    vector<int> count(range, 0);

    for (int i = 0; i < arr_size; ++i) {
        count[arr[i] - minVal]++;
        ops++;
    }

    int cntr = 0;
    for(int i = minVal; i <= maxVal; i++){
        for (int j = 0; j < count[i - minVal] ; j++) {
            arr[cntr] = i ;
            cntr++;
            ops++;
        }
    } 

    return ops;
}

int main() {
    vector<int> arr = {5, 12, 3, 4, 7, 1, 0, 6, 19, 8, 13, 4, 2, 10, 16};

    cout << "input: ";
    for(int num: arr){
        cout << num << " ";
    }    
    cout << endl;

    int ops = countingSort(arr);
    cout << "No of operations: " << ops << endl;

    cout << "Output: ";
    for (int i = 0; i < arr.size(); ++i) {
        cout << arr[i] << " ";
    }
    cout << endl;

    return 0;
}
